client/astarte_interface.go

```diff
@@ -0,0 +1,285 @@
+// Copyright © 2019 Ispirata Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+// AstarteInterfaceType represents which kind of Astarte interface the object represents
+type AstarteInterfaceType int
+
+const (
+	// PropertiesType represents a properties Interface
+	PropertiesType AstarteInterfaceType = iota
+	// DatastreamType represents a datastream Interface
+	DatastreamType
+)
+
+func (s AstarteInterfaceType) String() string {
+	return astarteInterfaceTypeToString[s]
+}
+
+var astarteInterfaceTypeToString = map[AstarteInterfaceType]string{
+	PropertiesType: "properties",
+	DatastreamType: "datastream",
+}
+
+var astarteInterfaceTypeToID = map[string]AstarteInterfaceType{
+	"properties": PropertiesType,
+	"datastream": DatastreamType,
+}
+
+// MarshalJSON marshals the enum as a quoted json string
+func (s AstarteInterfaceType) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString(`"`)
+	buffer.WriteString(astarteInterfaceTypeToString[s])
```

Comment:
You should probably rely on the existing string Marshaling and just do
```
json.Marshal(astarteInterfaceAggregationToString[s])
```
(There are a number of these in the PR, you should change them all)